snuba/datasets/groups.py

```diff
@@ -25,6 +27,25 @@
 from snuba.util import qualified_column
 
 
+class JoinedStorage(Storage):
```

Comment:
This, in this PR defies entirely the design (one storage per concrete table/view on clickhouse). It is removed in PR #830 so please see that one to have a good picture.
I did not support joins properly here because of the size of the PR.

Fundamentally the whole StorageQueryPlanBuilder abstraction (while it comes handy for the split queries and so on) is really needed datasets that include joins.
The reason this is needed is that we will have a query, that when ran on clickhouse, involves multiple Storages (like events and messages), and that, being the storage reusable, the specification on how to join them, is not in the storage, that comes from something at a higher level.
So we cannot get away with a simple Storage selection (we pick one storage over a list) in order to run a query like for discover.

In a world with Entities I see the query with joins working this way
1) The query defines which entities we want in the response (either explicitly saying which relationship we are interested into or implicitly. That is a query language matter, still we can say from the query which entities are invovled)
2) the dataset data model defines which relationship between entities are valid, thus which queries are valid.
3) we do dataset query processing as usual
4) we end that phase with an optimized join structure that includes only the entities we need to query to respond to the users.
5) This join structure is either in the SnubaQuery object (that does not exists) or aside. Which one we go with does not matter much here, the process would work the same way.
6) We feed that SnubaQuery to the JoinStorageQueryPlanBuilder of PR #830, and that one selects the Storages we need to join and transforms the query into a JoinClause which is a RelationalSource and that can be used in Storage query processing and Clickhouse Query Formatting. 
7) we do storage query processing and run the query.

Reply:
Ok, not reviewing this file per comment. The big picture here makes sense to me.

Reply:
Should this extend `ReadableStorage`?

Reply:
Technically, but this class disappears two PRs above this one. Will fix for the time being.